Draft: Feat/initial setup (#1)
StichtingOpenGeo/docs: src/tutorials/verbinden-met-zeromq.md

        print("Error bij decoderen")
        raise
        pass
    except:
        print("Error bij decomprimeren")
        raise
        pass

# Sluit de verbinding netjes af
subscriber.close()
context.term()
```

5. Vervang `{SERVER}` door de gewenste ZeroMQ server:

| Datafeed                | Adres                                     | Doel                                                     |
|-------------------------|-------------------------------------------|----------------------------------------------------------|
| BISON KV6 / KV15 / KV17 | tcp://pubsub.besteffort.ndovloket.nl:7658 | Verschillende bus-tram-metro data                        |
| BISON KV7/8 Turbo       | tcp://pubsub.besteffort.ndovloket.nl:7817 | Geplande en actuele reisinformatie bus-tram-metro        |
| NS InfoPlus             | tcp://pubsub.besteffort.ndovloket.nl:7664 | Verschillende trein data                                 |
| SIRI                    | tcp://pubsub.besteffort.ndovloket.nl:7666 | Nieuwe standaard voor geplande en actuele reisinformatie |

> Voor gebruikers die deze databronnen in productie willen gebruiken is een aparte versie beschikbaar met een SLA.

> > Neem contact op met het loket om hier gratis toegang tot te krijgen.

Reviewer: ```suggestion
> Neem contact op met het loket om hier toegang tot te krijgen.
```

> > **Let op**: voor gratis gebruikers is maximaal *één* actieve verbinding per datafeed toegestaan.

Reviewer: ```suggestion
> **Let op**: voor NDOV gebruikers is maximaal *één* actieve verbinding per datafeed toegestaan.
```

> Om meerdere projecten te ontsluiten kunnen datafeeds lokaal herdistribueert worden via [universal-sub-pubsub](https://github.com/StichtingOpenGeo/universal).

6. Start het Python script:

```
uv run main.py
```

Als de verbinding succesvol is ontvang je nu alle berichten op de gekozen ZeroMQ server.

> Krijg je errors in je terminal? In onze [Discord](https://discord.gg/gGjdt7uWQ) kunnen we je vast verder helpen.

Reviewer: ```suggestion
Krijg je fouten in je terminal? In onze [Discord](https://discord.gg/gGjdt7uWQ) kunnen we je vast verder helpen.
```

> # Verbinden met ZeroMQ

Reviewer: Schrijf hier eerst wat over `universal-sub-pubsub` zodat je al je software lokaal verbindt en er maar een verbinding naar NDOV loopt inclusief verbindingmanagement daarvoor.

> De meeste datafeeds van het NDOV Loket zijn ontsloten via het [ZeroMQ](https://zeromq.org/) protocol. Dit protocol maakt het mogelijk om een PubSub verbinding op te zetten, zodat je data automatisch gepusht krijgt zonder constant zelf nieuwe data op te hoeven halen. Voor deze basis tutorial zullen we gebruik maken van de programmeertaal Python met de [PyZMQ](https://github.com/zeromq/pyzmq) library.

Reviewer: ```suggestion
De meeste datafeeds van het NDOV Loket zijn ontsloten via het [ZeroMQ](https://zeromq.org/) protocol.
Dit protocol maakt het mogelijk om een PubSub verbinding op te zetten, zodat je data automatisch gepusht krijgt zonder constant zelf nieuwe data op te hoeven halen.
Voor deze basis tutorial zullen we gebruik maken van de programmeertaal Python met de [PyZMQ](https://github.com/zeromq/pyzmq) library.
```

```

4. Open `main.py` in een code editor naar keuze. Vervang de standaard code met onderstaande code-snippet:

```python
#!/usr/bin/env python3

# Imports
from gzip import GzipFile
from io import BytesIO
import time
import zmq

# Maak de nieuwe subscriber socket aan
context = zmq.Context()
subscriber = context.socket(zmq.SUB)

# Verbind met NDOV Loket
subscriber.connect("{SERVER}")

# Vertel de server welke data je allemaal wil ontvangen.
# `/` is een wildcard die er voor zorgt dat je alle data ontvangt.
subscriber.setsockopt_string(zmq.SUBSCRIBE, "/")

# Ontvang alle data en loop door de nieuwe berichten
while True:
    multipart = subscriber.recv_multipart()

    # Berichten ontstaan uit 2 delen, het adres en de inhoud
    address = multipart[0].decode('UTF-8')

    # Om dataverkeer te besparen wordt alle inhoud gecomprimeerd met gzip,
    # dit moeten we dus handmatig decomprimeren
    try:
        contents = b''.join(multipart[1:])
        contents = GzipFile('','r',0,BytesIO(contents)).read()
        contents = contents.decode('UTF-8')

        # Print de tijd, het adres en de inhoud naar je terminal
        print(int(time.time()), address, contents)

    except:

>         print("Error bij decomprimeren")

Reviewer: Dat zou ook og een UTF-8 error kunnen zijn, misschien die Except specifieker maken?

        raise
        pass

# Sluit de verbinding netjes af
subscriber.close()
context.term()
```

5. Vervang `{SERVER}` door de gewenste ZeroMQ server:

| Datafeed                | Adres                                     | Doel                                              |
|-------------------------|-------------------------------------------|---------------------------------------------------|
| BISON KV6 / KV15 / KV17 | tcp://pubsub.besteffort.ndovloket.nl:7658 | Verschillende bus-tram-metro data                 |
| BISON KV7/8 Turbo       | tcp://pubsub.besteffort.ndovloket.nl:7817 | Geplande en actuele reisinformatie bus-tram-metro |
| NS InfoPlus             | tcp://pubsub.besteffort.ndovloket.nl:7664 | Verschillende trein data                          |

> | SIRI                    | tcp://pubsub.besteffort.ndovloket.nl:7666 | ?                                                 |

Reviewer: ```suggestion
| SIRI                    | tcp://pubsub.besteffort.ndovloket.nl:7666 | Nieuwe standaard voor geplande en actuele reisinformatie |
```

> > Voor gebruikers die deze databronnen in productie willen gebruiken is een aparte versie beschikbaar met een SLA. Neem contact op met het loket om hier gratis toegang tot te krijgen.

Reviewer: ```suggestion
> Voor gebruikers die deze databronnen in productie willen gebruiken is een aparte versie beschikbaar met een SLA.
> Neem contact op met het loket om hier toegang toe te krijgen.
```

6. Start het Python script:

```
uv run main.py
```

Als de verbinding succesvol is ontvang je nu alle berichten op de gekozen ZeroMQ server.

> Krijg je errors in je terminal? In onze [Discord]() kunnen we je vast verder helpen.

Reviewer: ```suggestion
Krijg je fouten in je terminal? In onze [Discord](https://discord.gg/gGjdt7uWQt) kunnen we je vast verder helpen.
```
